### PR TITLE
improve buffer handling inside SslStream

### LINKED ISSUE #49000: SSlStream reads TLS records one frame at a time
### Description

See https://github.com/dotnet/aspnetcore/issues/30545#issuecomment-788557331 for more background on the issue. Kestrel does 4K buffer writes to SslStream on the server side and this results in only being able to read 4K buffers on the client side. The amount of TLS frames shouldn't affect how they are read on the client side. The extra cost here ends up being the amount of reads that need to be performed client side to quickly read data (paying for lots of async machinery)

### Configuration

I tested on Windows, I haven't tried on linux but haven't found any code that makes me believe this is a windows specific problem.

### Regression?

Nope, seems to happen on .NET Core 3.1 as well.

### Analysis

The loop here ends as soon as we have decrypted *any* data https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L766. Though this isn't a stream contract requirement, the loop should end in 3 cases:

1. We don't have another TLS frame and haven't seen EOF.
2. We have seen EOF
3. We have run out of space to fill the user's buffer.

### Sample Program

```C#
using System;
using System.Buffers;
using System.IO;
using System.IO.Pipelines;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Threading;
using System.Threading.Tasks;

var pair = DuplexPipe.CreateConnectionPair(new PipeOptions(), new PipeOptions());

var server = new SslStream(new DuplexPipeStream(pair.Application.Input, pair.Application.Output));
var client = new SslStream(new DuplexPipeStream(pair.Transport.Input, pair.Transport.Output));

async Task Server()
{
    var certString = "MIIJrwIBAzCCCWsGCSqGSIb3DQEHAaCCCVwEgglYMIIJVDCCBfoGCSqGSIb3DQEHAaCCBesEggXnMIIF4zCCBd8GCyqGSIb3DQEMCgECoIIE/jCCBPowHAYKKoZIhvcNAQwBAzAOBAiaFr6Goa0S+gICB9AEggTY3sKr0dEZ+MINnW2jAcY2aSJTv1pyzQdBtUIEJO38ygeoTAhNCJcXGiyQGu8WZXid5wxgjs6gztSfK/chFvOq9O3ZG378ygBssFcY4vjFYwFddg6seMGMlxm/su6PPCvnRWX/YQ/YOgsHeEKVoHiLLPze05WRzNlt1OPjvmrgQUhY0drVXGdXXxqfPP/f+rMluCOA6qZUJqrMLQR4Eso9hyJhCi9ubQq61HTR+ZDFWjM/yofXe6y3SpwdCrr5zKz7Uu2BVq8R5zO50tjAHwo8znJkLmnrEN2rDUMGkAHMPyt4J4TwPtTxR7Ilk0+An6eriIBsx/tIDZnrlLva7qVHEh6z9LdPCa5u/4qFvisbkvJ+T68C7BX3e6TWLL7VZ4rbVMMlnv7YBebpDkQvrVoZ5V3GLsgKMY6a1ivlj9QFlZs7Oea0xyeFMMX4fh5N6j8Ap9nsySEbJZ6Kw6PuVqfwNfr3TjLlPHCDNPHv5q/Xmfdisgq9w1M1HLP9xkYrBhUu7pV/IklBfdGn4A8okI5AeAuF9Aumugq3IyOc3hpxwHp9T7821xDbYejXHGI30Ehg6N3ASl71bm99OXrh80NHt2Ogqo2OyeUEBIECIbnnEluEPsm3Lsgq5NP81VFOVdTbS/nhYMYLPJVNQIARMam00WWcx6HgpaR3NZkw1bR9njbA4MRPFkMmZRbEevN2/9myoCSN796/JO6V9itH18bbHA5JS1l5mRA0xarjaS8mz9J8fX2JOrxN5U/yBjZkrt/VxEkzmlFQvnpsOp63cnnEsgrjw2MC/7uZ2tRNpcBON3cjqPDpxoiyG+kOq/aAu0p7XJGkGbiXUyjkvwKZUbVktBETU7199oj2zPU9h2wcQ52NeiWEa8b5FveCjXwRz5kUp0VzbBlw1/4STyANZ/YQ6CoQ4bm/bLb+p8/SrjheTe2D2FDE19ZsEPzHFWHFYFlfKlOMed6+rZZnKCHT6wijRHyex9C4otOiQ9cDhj70x87sAWnTN9nZbcaczl7g3QsmOdRjLBqnrh5cQCOrPhDPPCwR0CocH035ASnoJByd4WiZeW19Bz27K78tuT7vlOT6fHyf4EQjhE/4YsgCmBKIwS6upRECheYo8FG8Gg5eY8mjk9p3i9DHjKxwh/de3U/s/0HEEz3Vj75yqTs6NjQYiD1BckKevm2B9Pq3Lb1ZpVOULJddiiMlWGu37nmzOTA1968CxkMJhvpPDvTwOpOnnv83H3xs6iJ7f7rV8jerCQxelNuRfa5E26XUbAP2/HkE5XDW/oA0e4f74fqLvyTUrw3IZwSgQN9Y0/FDrAHIVNpxiZ5LxjEg9EnVhfBh1vPw5LwcCgUuQ++rhSoW9WD5pygp8R1alDWprxbJmOclHt5uIcS7JHih8sonPXjUSXe9BWdSFcxsfBRJo8vRfeUj8Jt3x7HeqfoSDaAhSdZ3jtthkschDzMbI8KGKXZLgDaD3CJo+MZNe+cg5zmTfhYpHjKJW41P6XNnJLSxV5YfugLF9vE+JUEea17bOTiwxyTLqrCv3UuaW2YxDpK+5/i2hYalJJ0kwLlX0CJmcScpyp9dkgnPl9mwlt4exxBHY5WxMMq7f4z454CpmpJmuvjJCTDbS/dGGb+/DLxp7k51m3VJ1B40MeIweDGBzTATBgkqhkiG9w0BCRUxBgQEAQAAADBXBgkqhkiG9w0BCRQxSh5IADQAYgBhADMAOABlADAAMwAtAGEAOQBhAGMALQA0ADMAMAA3AC0AOAA5ADcAYQAtADQAMwBhADMAYwA0ADgANQBhADQANwA4MF0GCSsGAQQBgjcRATFQHk4ATQBpAGMAcgBvAHMAbwBmAHQAIABTAHQAcgBvAG4AZwAgAEMAcgB5AHAAdABvAGcAcgBhAHAAaABpAGMAIABQAHIAbwB2AGkAZABlAHIwggNSBgkqhkiG9w0BBwGgggNDBIIDPzCCAzswggM3BgsqhkiG9w0BDAoBA6CCAw8wggMLBgoqhkiG9w0BCRYBoIIC+wSCAvcwggLzMIIB36ADAgECAhAytp0sn8UPr08doiEdX3DhMAkGBSsOAwIdBQAwFDESMBAGA1UEAxMJbG9jYWxob3N0MB4XDTE1MDkxNjA2MjEwOVoXDTE2MDkxNjA2MjEwOFowFDESMBAGA1UEAxMJbG9jYWxob3N0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAmlGZoaDWCOXZWDXlwaUJjGvfdpUpxYdoGxiHBiUowr/l+fjFq+xpqgjS/5VSXd64qq3ErofOc7yUrTFANMdZG6dVfzRaoHNXOZQjmvbGMOOOJQw4zc/Bt5DTqyeRt462gxLP/HkIE45NE8qRvsH6uU1BN1q/pHxsnkaeURqFcNgVR1f7BQfobj1hpneEBIYh75GRpSUo02VS5gGDc+B/sWB2fns7m05rnquYmpAQRQ7NyvoVn3JYj/EFN7RpdltqDytveMV50SSqLnKbcEmzBlRZWLXfdFNJjIwMBvI4z9nGkJFMvmn0qFAuPH+5WEZXN5F6OU83EQ0oAURfnJkEcwIDAQABo0kwRzBFBgNVHQEEPjA8gBAZzi0Ai4TG1VvIcjfkv3KIoRYwFDESMBAGA1UEAxMJbG9jYWxob3N0ghAytp0sn8UPr08doiEdX3DhMAkGBSsOAwIdBQADggEBAHGylUp5KGHHIlfAdmgcGZ89xtE5hpP0Wsd8KlE+HKfIDbt5SB7YBGl8gEMUAGMnVvBkRHtLmTJO0Ez3VOmTLCAOywOkduFXNsOjVkwPyCUnLlz7EriNpNmPT9ZWTVQdFV7FRXhjEgRXLCX+tHz1IF677MIx6kL47AmVKnr+9g+Fr5BfZDkNZpapJK3mflIdaM14jOz6rsqsWsJMCOlQPKcltq7BTKgIquNPDehHyFQjbMvMZg9Kf/YbpuJKFpCzfi+uiMyc5StBRVIF2buKQnvCiuTe6HUE6ZcrJxsEx7dSJ/lG5h3gPsfoNWxnTiomWV/Zp2E2YwfavjPhWjrBwrIxFTATBgkqhkiG9w0BCRUxBgQEAQAAADA7MB8wBwYFKw4DAhoEFK1iGPraW4Gf5vcA8VhmKYAjnMYSBBQ3NJrH26kpqUt0uX/tsyJX2WCqKAICB9A=";
    var cert = new X509Certificate2(Convert.FromBase64String(certString), "testPassword");
    await server.AuthenticateAsServerAsync(cert);

    // Write 10500, 100 byte frames
    for (int i = 0; i < 10500; i++)
    {
        await server.WriteAsync(new byte[100]);
    }
}

async Task Client()
{
    await client.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
    {
        TargetHost = "localhost",
        RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true
    });

    // 1MB
    var buffer = new byte[0x100000];

    do
    {
        var read = await client.ReadAsync(buffer).ConfigureAwait(false);
        if (read == 0)
        {
            break;
        }

        Console.WriteLine("Read: " + read);
    }
    while (true);
}

var s1 = Server();
var s2 = Client();
await Task.WhenAll(s1, s2);


internal class DuplexPipe : IDuplexPipe
{
    public DuplexPipe(PipeReader reader, PipeWriter writer)
    {
        Input = reader;
        Output = writer;
    }

    public PipeReader Input { get; }

    public PipeWriter Output { get; }

    public static DuplexPipePair CreateConnectionPair(PipeOptions inputOptions, PipeOptions outputOptions)
    {
        var input = new Pipe(inputOptions);
        var output = new Pipe(outputOptions);

        var transportToApplication = new DuplexPipe(output.Reader, input.Writer);
        var applicationToTransport = new DuplexPipe(input.Reader, output.Writer);

        return new DuplexPipePair(applicationToTransport, transportToApplication);
    }

    // This class exists to work around issues with value tuple on .NET Framework
    public readonly struct DuplexPipePair
    {
        public IDuplexPipe Transport { get; }
        public IDuplexPipe Application { get; }

        public DuplexPipePair(IDuplexPipe transport, IDuplexPipe application)
        {
            Transport = transport;
            Application = application;
        }
    }
}

internal class DuplexPipeStream : Stream
{
    private readonly PipeReader _input;
    private readonly PipeWriter _output;
    private readonly bool _throwOnCancelled;
    private volatile bool _cancelCalled;

    public DuplexPipeStream(PipeReader input, PipeWriter output, bool throwOnCancelled = false)
    {
        _input = input;
        _output = output;
        _throwOnCancelled = throwOnCancelled;
    }

    public void CancelPendingRead()
    {
        _cancelCalled = true;
        _input.CancelPendingRead();
    }

    public override bool CanRead => true;

    public override bool CanSeek => false;

    public override bool CanWrite => true;

    public override long Length
    {
        get
        {
            throw new NotSupportedException();
        }
    }

    public override long Position
    {
        get
        {
            throw new NotSupportedException();
        }
        set
        {
            throw new NotSupportedException();
        }
    }

    public override long Seek(long offset, SeekOrigin origin)
    {
        throw new NotSupportedException();
    }

    public override void SetLength(long value)
    {
        throw new NotSupportedException();
    }

    public override int Read(byte[] buffer, int offset, int count)
    {
        // ValueTask uses .GetAwaiter().GetResult() if necessary
        // https://github.com/dotnet/corefx/blob/f9da3b4af08214764a51b2331f3595ffaf162abe/src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs#L156
        return ReadAsyncInternal(new Memory<byte>(buffer, offset, count), default).Result;
    }

    public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken = default)
    {
        return ReadAsyncInternal(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
    }

    public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
    {
        return ReadAsyncInternal(destination, cancellationToken);
    }

    public override void Write(byte[] buffer, int offset, int count)
    {
        WriteAsync(buffer, offset, count).GetAwaiter().GetResult();
    }

    public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
    {
        if (buffer != null)
        {
            _output.Write(new ReadOnlySpan<byte>(buffer, offset, count));
        }

        await _output.FlushAsync(cancellationToken);
    }

    public override async ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
    {
        _output.Write(source.Span);
        await _output.FlushAsync(cancellationToken);
    }

    public override void Flush()
    {
        FlushAsync(CancellationToken.None).GetAwaiter().GetResult();
    }

    public override Task FlushAsync(CancellationToken cancellationToken)
    {
        return WriteAsync(null, 0, 0, cancellationToken);
    }

    private async ValueTask<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken)
    {
        while (true)
        {
            var result = await _input.ReadAsync(cancellationToken);
            var readableBuffer = result.Buffer;
            try
            {
                if (_throwOnCancelled && result.IsCanceled && _cancelCalled)
                {
                    // Reset the bool
                    _cancelCalled = false;
                    throw new OperationCanceledException();
                }

                if (!readableBuffer.IsEmpty)
                {
                    // buffer.Count is int
                    var count = (int)Math.Min(readableBuffer.Length, destination.Length);
                    readableBuffer = readableBuffer.Slice(0, count);
                    readableBuffer.CopyTo(destination.Span);
                    return count;
                }

                if (result.IsCompleted)
                {
                    return 0;
                }
            }
            finally
            {
                _input.AdvanceTo(readableBuffer.End, readableBuffer.End);
            }
        }
    }

    public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
    {
        throw new NotSupportedException();
    }

    public override int EndRead(IAsyncResult asyncResult)
    {
        throw new NotSupportedException();
    }

    public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
    {
        throw new NotSupportedException();
    }

    public override void EndWrite(IAsyncResult asyncResult)
    {
        throw new NotSupportedException();
    }
}
```

### LINKED ISSUE #49086: SslStream should avoid copying read buffer data when it already has another frame to process
The logic in SslStream.ReadAsyncInternal calls ResetReadBuffer whenever it finishes processing the previous decrypted frame and thus needs to decrypt another frame. However, this is unnecessarily aggressive. We should only do this when we are about to actually issue a read against the underlying stream. If we already have a frame buffered, then we won't read and will just process that frame immediately, which means the copy was pointless.

This might be a good reason to use ArrayBuffer here, since it simplifies some of the buffer management.

Related to #49000 
Related to #49019

### PR DESCRIPTION
**Pull request by wfurt**
This is followup-up on discussion in #49000
This really have 2 separate change + some cleanup. I may be able to do them independently but they are somewhat intertwined so I did them and test them together. 

1. Is issue discussed in #49000 to process multiple TLS frames at once when already present in internal buffer. 
Instead of doing it in PAL as original prototype this does it in ReadAsyncInternal() in platform independent (mostly) way. 
This had two basic challenges to that complicated the matter.

- the decrypted data are _always_ smaller (or frame may produce 0 bytes) so when decrypted, they would land in disjoined fragments. using single offset and sound become problematic. To solve this, I turned the input to separate input `Span<byte>`. That also avoids several checks and calculations for the offset. It also makes the code less confusing IMHO as we no longer use "output" variable as input. In general, once we read the frames there is no longer async boundary and using Span as much as possible is nice and easy. 

- While processing the extra frames we may reach end of stream. Either by actually reaching EOF on underlying stream or by reading end processing TLS close notification. To solve this I added extra bool so we can return previous data and we return EOF in next round. 

2. We can avoid allocating/renting 16k internal buffer in some cases. As part of the spanification above, I turned _internalBuffer into `Memory<byte>`. That either points to rented buffer or it can point to user provided buffer if that is large enough. That has three benefits:
- we can reduce memory utilization. Even if rented, this can add up in aggregate use. We may need to allocate the buffer anyway if we end up with partial frames but that is no worse than now and we need to copy only the leftover instead of everything. 
- On Linux & macOS we can write decrypted data directly to the right location without any extra copy. That is not possible on Windows as Schannel always decrypt in place.
- We can possibly do much bigger reads from underlying stream. Right now we are limited to 16K + what ever we get extra from rented buffer. Loopback on Linux is 64K so we may be be able to do much bigger reads. That is also true for TCP with large Window and fast network. When we use the provided buffer, user can indicate how much there are willing to trade. The example @davidfowl provided, I can see now 50K reads with no extract cost. (and no new API)

This starts in ResetReadBuffer() but with the Memory<byte>, the impact on rest of the code is pretty small. 

3. I kick few items out of the main processing loop. Processing already decrypted data should be no brainer. I also moved the zero byte processing. That gives caller to provide real buffer and avoid copy in some cases. I can possibly limit that with _internalBufferCount == 0 e.g. proceed to normal read if we have some data and buffer already allocated. 

Testing:

The SchSendAux* test do depend on particular read size and they take it guess that the AUX record is present - Windows only. That seems generally problematic and since I did not find any good way how to do better I deleted the tests. I feel that is something the PAL is responsible for and I don't see big value in the tests. 

Running the existing tests (SSL & HTTP) I found it difficult to hit some of the code branches. As produces write to SslStream we write it to Memory or socket and that shows up as whole on the other side. I added RandomIOStream that creates short reads and writes e.g. reads less then asked for and splits writes to multiple writes to underlying stream. That helped me to uncover interesting cases I may have missed otherwise. 

I still plan to run more benchmarks but I would like to get some feedback before that is all completed. 

fixes #49000
fixes #49086

contributes to (closed) #49019

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ private enum Framing
         private const int ReadBufferSize = 4096 * 4 + FrameOverhead;         // We read in 16K chunks + headers.
         private const int InitialHandshakeBufferSize = 4096 + FrameOverhead; // try to fit at least 4K ServerCertificate
         private ArrayBuffer _handshakeBuffer;
+        private bool receivedEOF;
```
Comment (davidfowl) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs:44]: nit: _receivedEOF

**Comment by benaadams** [comment]
Giving compile errors "Field 'SslStream._rentedBuffer' is never assigned to, and will always have its default value null"

**Comment by wfurt** [comment]
I'm looking. _rentedBuffer is assigned in ResetReadBuffer().

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 2]
```diff
@@ -427,6 +429,90 @@ public async Task SslStream_ClientCertificate_SendsChain()
...
+
+                        data = data.Slice(writeBufferSize);
+                    }
+
+                    server.ShutdownAsync().GetAwaiter().GetResult();
+                });
+
+                Task reader = Task.Run(() =>
+                {
+                    SHA256 hash = SHA256.Create();
```
Comment (danmoseley) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:480]:
```suggestion
                    using SHA256 hash = SHA256.Create();
```

Comment (davidfowl) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:480]: I thought we had static's now

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 3]
```diff
@@ -436,5 +522,97 @@ public async Task SslStream_ClientCertificate_SendsChain()
...
+                get
+                {
+                    return _innerStream.Position;
+                }
+
+                set
+                {
+                    _innerStream.Position = value;
+                }
+            }
```
Comment (danmoseley) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:556]:
```suggestion
            public override long Position
            {
                get => _innerStream.Position;
                set => _innerStream.Position = value;
            }
```

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 4]
```diff
@@ -427,6 +429,90 @@ public async Task SslStream_ClientCertificate_SendsChain()
...
 
+        [Theory]
+        [InlineData(16384 * 100, 4096, 1024, true)]
+        [InlineData(16384 * 100, 1024 * 20, 1024, true)]
+        public async Task SslStream_RandomWrites_OK(int bufferSize, int readBufferSize, int writeBufferSize, bool useAsync)
+        {
+            byte[] dataToCopy = RandomNumberGenerator.GetBytes(bufferSize);
+            byte[] srcHash = SHA256.Create().ComputeHash(dataToCopy);
+
+            var clientOptions = new SslClientAuthenticationOptions() { TargetHost = "localhost", };
```
Comment (danmoseley) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:440]:
```suggestion
            var clientOptions = new SslClientAuthenticationOptions() { TargetHost = "localhost" };
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 5]
```diff
@@ -58,11 +58,11 @@ public partial class SslStream : AuthenticatedStream
...
         // Never updated directly, special properties are used.  This is the read buffer.
-        internal byte[]? _internalBuffer;
-        internal int _internalOffset;
-        internal int _internalBufferCount;
-        internal int _decryptedBytesOffset;
-        internal int _decryptedBytesCount;
+        internal int _internalOffset;               // This is offset where encrypted data starts.
+        internal int _internalBufferCount;          // Length of encrypted data.
+        internal int _decryptedBytesOffset;         // Offset of decrypted data relative to _internalBuffer.
+        internal int _decryptedBytesCount;          // Number of already decrypted bytes. Really valid only ion combination with _rentedBuffer.
```
Comment (danmoseley) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.cs:64]:
```suggestion
        internal int _decryptedBytesCount;          // Number of already decrypted bytes. Really valid only in combination with _rentedBuffer.
```

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:480]: Ah yes, you can just do `SHA256.HashData(...)`. No Create() and no dispose.

Comment (wfurt) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:480]: That works only when you have whole chunk. Here we feed `TransformFinalBlock` one chunk at the time.

**Comment by geoffkizer** [comment]
It feels to me like this PR is mixing together several changes that we should instead implement and measure independently.

In particular:
(1) Process multiple TLS frames in Read (#49000) -- This has little downside that I can see and we should just do it.
(2) Avoid extra copy of decrypted data on Linux/MacOS -- This seems like goodness, but it complicates the logic quite a bit here as it now works very differently for Linux vs Windows, and not just at the PAL layer. (In fact the PAL layer now works very differently for Linux and Windows, and that seems bad.) I think this would be nice to do, but I'd like to (a) do it in a way that doesn't introduce as much complexity, and (b) measure that there is actually a win here.
(3) Use user-provided buffer instead of internal rented buffer when possible. Again, this seems like goodness, but it also seems to introduce a lot of complexity. It's also not clear to me how much of a win this is, since in practice we are going to end up allocating the internal buffer in a lot of cases anyway. 

I think it would be better to tackle each one of these independently (starting with (1)), validate the win in benchmarks or user scenarios etc, rinse and repeat.

Thoughts?

Also:
- Using Span is definitely good, whether in the PAL or elsewhere -- this would have avoided that other issue we found recently
- I'm fine deleting the SchSendAux tests
- The RandomStream stuff looks cool!

**Comment by geoffkizer** [comment]
In particular for (2):

I think we could have two functions at the PAL layer -- something like DecryptToBuffer and DecryptInPlace. 

On Linux, DecryptInPlace would just call DecryptToBuffer with the original buffer.
On Windows, we'd always decrypt in place (since we have to) and then DecryptToBuffer would do the copy to the target buffer.

Thoughts?

**Comment by geoffkizer** [comment]
I also wonder how some of these (particularly 2 and 3) would relate to implementing a custom BIO, like we have discussed in the past. Is this something we are still considering? How would we prioritize it relative to other stuff here?

**Comment by wfurt** [comment]
macOS for example does not use BIO. (and we will need to re-write it when to go for ALPN & Tls13) 
We may consider it for Linux but AFAIK the perf improvement was small when tried last time. 

The 2 & 3 is essentially same code. For sake of this PR I can roll that back. We could split the Decrypt*()  into two parts but Windows _always_ decrypt in place and Unix _always_ copy. I was originally thinking about hiding the difference in Windows PAL but we may end up with unnecessary copy. I'll think about it more and we can discuss details on the PR.

**Comment by geoffkizer** [comment]
> macOS for example does not use BIO. (and we will need to re-write it when to go for ALPN & Tls13)

Yeah, but we don't really care about perf for macOS. I mean if we can get a win on macOS too, then great, but it's not a priority.

> We may consider it for Linux but AFAIK the perf improvement was small when tried last time.

As I understand it, we are copying all input data into the memory BIO today. A custom BIO would allow us to avoid this copy, and possibly have other benefits as well. It seems like the benefit of that would by definition be more significant than just (2) above. So if the BIO win is insignificant, then (2) is as well; conversely, if we think there may be a win here, then the BIO win seems like it would be larger than (2).

> The 2 & 3 is essentially same code. 

I don't understand this. (3) applies to both Linux and Windows, right? Whereas (2) is only a benefit for Linux.

> I was originally thinking about hiding the difference in Windows PAL but we may end up with unnecessary copy. I'll think about it more and we can discuss details on the PR.

Sounds good. In general we should strive to encapsulate the platform differences in the PAL. This is not always possible, but if it's not then at the least we need to expose different plat-specific functions from the PAL, as opposed to having one that looks like it's platform neutral but actually is not.

**Comment by alnikola** [comment]
I suspect it might be causing a perf regression in YARP HTTPS (HTTP 1.1) scenario. Let's investigate that before merging.

**Comment by davidfowl** [comment]
> I suspect it might be causing a perf regression in YARP HTTPS (HTTP 1.1) scenario. Let's investigate that before merging.

What's causing a regression?

**Comment by alnikola** [comment]
This PR. I tried running YARP with a privately built System.Net.Security.dll and saw a perf drop along with some bad responses in YARP HTTPS - HTTP (HTTP 1.1) scenario on Windows.

**Comment by davidfowl** [comment]
😢

**Comment by wfurt** [comment]
It is still coming but slow @davidfowl

**Comment by davidfowl** [comment]
@wfurt Can we revive this?

**Comment by wfurt** [comment]
yes, it is on my todo list for 8 @davidfowl. I think I need better plan for Windows.

**Comment by davidfowl** [comment]
Is there an issue tracking it?

**Pull request closed without merging**